### PR TITLE
SNAP-656 Delink RDD partitions from buckets 

### PR DESCRIPTION
**Pull request by ymahajan**
## Changes proposed in this pull request
- set default rdd partitions as number of cores in cluster 
- map buckets to target partitions using round robin assignment
- colocated tables should have same partitions and queries should work fine. 
- split mode support 

## ReleaseNotes.txt changes
yes

## Other PRs 
Store - https://github.com/SnappyDataInc/snappy-store/pull/85
Spark - https://github.com/SnappyDataInc/spark/pull/4
SnppyData - https://github.com/SnappyDataInc/snappydata/pull/297

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala [hunk 1]
```diff
@@ -90,7 +92,15 @@ class BaseColumnFormatRelation(
   @transient protected lazy val region = Misc.getRegionForTable(resolvedName,
     true).asInstanceOf[PartitionedRegion]
 
-  override lazy val numPartitions: Int = region.getTotalNumberOfBuckets
+  override lazy val numPartitions: Int = {
+//    val region = Misc.getRegionForTable(resolvedName, true).
+//        asInstanceOf[PartitionedRegion]
+//    region.getTotalNumberOfBuckets
+    val numCores = Runtime.getRuntime.availableProcessors()
```
Comment (rishitesh) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala]: This is evaluated at driver node. We need to consider sever nodes. Driver node num processors is not useful to us. Can you please see SchedulerBackend.defaultParallelism. That  takes total cores for slaves into consideration.
Catch however is spark.default.parallelism gets priority and if somebody configures bad we will suffer.

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala [hunk 2]
```diff
@@ -339,6 +349,8 @@ class BaseColumnFormatRelation(
         val sql = s"CREATE TABLE $tableName $schemaExtensions " + " DISABLE CONCURRENCY CHECKS "
         logInfo(s"Applying DDL (url=${connProperties.url}; " +
             s"props=${connProperties.connProps}): $sql")
+        println(s"Applying DDL (url=${connProperties.url}; " +
```
Comment (rishitesh) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala]: Remove println

File: core/src/main/scala/org/apache/spark/sql/execution/row/RowFormatRelation.scala [hunk 3]
```diff
@@ -138,11 +139,20 @@ class RowFormatRelation(
...
-      case pr: PartitionedRegion => pr.getTotalNumberOfBuckets
-      case _ => 1
+      case pr: PartitionedRegion =>
+        getNumPartitions
+      case _ =>
+        1
     }
   }
 
+  def getNumPartitions : Int = {
```
Comment (rishitesh) [inline core/src/main/scala/org/apache/spark/sql/execution/row/RowFormatRelation.scala]: See comments above

File: core/src/test/scala/org/apache/spark/sql/NorthWindTest.scala [hunk 4]
```diff
@@ -0,0 +1,1481 @@
...
+package org.apache.spark.sql
+
+import io.snappydata.SnappyFunSuite
+import org.apache.spark.Logging
+import org.apache.spark.sql.execution.aggregate.{SortBasedAggregate, TungstenAggregate}
+import org.apache.spark.sql.execution.joins._
+import org.apache.spark.sql.execution.{LocalTableScan, PartitionedPhysicalRDD, Project, Sort}
+import org.scalatest.{BeforeAndAfter, BeforeAndAfterAll}
+
+class NorthWindTest
```
Comment (rishitesh) [inline core/src/test/scala/org/apache/spark/sql/NorthWindTest.scala]: Remove customer name from class.

File: core/src/test/scala/org/apache/spark/sql/NorthWindTest.scala [hunk 5]
```diff
@@ -0,0 +1,1481 @@
...
+    assert(df.rdd.partitions.length == numPartitions,
+      "Mismatch got df.rdd.partitions.length ->" + df.rdd.partitions.length +
+        " but expected numPartitions ->" + numPartitions)
+  }
+
+  private def createAndLoadColocatedTables(snc: SnappyContext): Unit = {
+
+    val regions = snc.read.format("com.databricks.spark.csv")
+      .option("header", "true")
+      .load(getClass.getResource("/northwind/regions.csv").getPath)
```
Comment (rishitesh) [inline core/src/test/scala/org/apache/spark/sql/NorthWindTest.scala]: remove specific names

File: core/src/main/scala/org/apache/spark/sql/store/StoreUtils.scala [hunk 6]
```diff
@@ -98,25 +101,74 @@ object StoreUtils extends Logging {
...
+      val bucketSet = {
+        if (serverToBuckets.contains(owner)) serverToBuckets.get(owner).get
+        else new mutable.HashSet[Int]()
+      }
+      bucketSet += p
+      totalBuckets += p
+      serverToBuckets put(owner, bucketSet)
+    }
+    val numCores = Runtime.getRuntime.availableProcessors()
+    val numServers = GemFireXDUtils.getGfxdAdvisor.adviseDataStores(null).size()
```
Comment (rishitesh) [inline core/src/main/scala/org/apache/spark/sql/store/StoreUtils.scala]: This part wont be straight forward. If executor core is not set I think default is all the cores of a server. If different configuration machines are used our calculation will go for toss. e.g. if a server is with 8 cores and another with 16 cores, then we will loose processing power of some of the servers.

File: tests/common/src/main/resources/northwind/employees.csv [hunk 7]
```diff
@@ -0,0 +1,10 @@
+EmployeeID,LastName,FirstName,Title,TitleOfCourtesy,BirthDate,HireDate,Address,City,Region,PostalCode,Country,HomePhone,Extension,Photo,Notes,ReportsTo,PhotoPath
```
Comment (rishitesh) [inline tests/common/src/main/resources/northwind/employees.csv]: Is it ok to use this data?

Comment (ymahajan) [inline tests/common/src/main/resources/northwind/employees.csv]: Yes, NorthWind Trader's schema is widely used in public domain, especially with the database research students. Also this data is shipped with Microsoft products.

File: core/src/test/scala/org/apache/spark/sql/NorthWindTest.scala (hunk 4 above)
Comment (ymahajan) [inline core/src/test/scala/org/apache/spark/sql/NorthWindTest.scala]: NorthWind Trader's is fictitious company name, not any real SnappyData customer's name.

File: core/src/main/scala/org/apache/spark/sql/store/StoreUtils.scala (hunk 6 above)
Comment (ymahajan) [inline core/src/main/scala/org/apache/spark/sql/store/StoreUtils.scala]: We have assumed that, the cluster will be homogeneous with respect number of cores on each server. I will check if there is a way to handle heterogeneous case also.

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala (hunk 2 above)
Comment (ymahajan) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala]: Removed println

**Pull request merged**